src/atomic.rs

```diff
@@ -0,0 +1,860 @@
+use std::borrow::{Borrow, BorrowMut};
+use std::marker::PhantomData;
+use std::mem;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use Scope;
+
+/// Given ordering for the success case in a compare-exchange operation, returns the strongest
+/// appropriate ordering for the failure case.
+#[inline]
+fn strongest_failure_ordering(ord: Ordering) -> Ordering {
```

Comment:
Regarding naming, here's my breakdown of possible names:

1. `compare_and_swap` - the "swap" here is misleading because the returned value on success is not the old one. That really isn't a swap. :)
2. `compare_exchange` - here, "exchange" has equivalent meaning to "swap". Again, it doesn't exchange values on success. Moreover, the behavior of `compare_exchange` in `libstd` is different.
3. `compare_and_set` - doesn't have any misleading connotations, as far as I can tell. Perhaps this is the most correct naming choice?

I like the `Into<Orderings>` idea, seems like a good solution. :) I've implemented something like it in the form of `CasOrdering` trait. Let me know what you think about that.